### PR TITLE
Java call Python use structured function descriptors

### PR DESCRIPTION
**Pull request by fyrestone**
## Why are these changes needed?

Java call Python APIs use string to describe Python functions. e.g. `Ray.callPy("moduleName", "functionName", arg0, ...)`. The description data is mixed with arguments.

So, this PR introduces three types for describing Python functions.
  - PyActorClass
  - PyActorMethod
  - PyRemoteFunction

The API names are also changed:
  - Ray.callPy -> Ray.call
  - Ray.createPyActor -> Ray.createActor

## Why not use `PythonFunctionDescriptor` directly?

PythonFunctionDescriptor is a protobuf message for cross language serialization. But, it has only one type contains all necessary fields, it's hard to determine when to fill which field. 

Besides, the three types has some fields are not necessary to serialize. e.g. returnType

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: java/api/src/main/java/org/ray/api/function/PyActorClass.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.ray.api.function;
+
+/**
+ * A class for creating Python actor.
+ */
+public class PyActorClass {
+  public String moduleName;
+  public String className;
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/function/PyActorClass.java]: `public final` to make it immutable.

File: java/api/src/main/java/org/ray/api/function/PyActorClass.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.ray.api.function;
+
+/**
+ * A class for creating Python actor.
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/function/PyActorClass.java]:
```suggestion
 * A class that represents a Python actor class.
```

File: java/api/src/main/java/org/ray/api/function/PyActorMethod.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.ray.api.function;
+
+/**
+ * A class for calling actor method.
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/function/PyActorMethod.java]:
```suggestion
 * A class that represents a method of a Python actor. 
 *
 * Note, information about the actor will be inferred from the actor handle, so it's not specified in this class.
```

File: java/api/src/main/java/org/ray/api/function/PyActorMethod.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+ */
+public class PyActorMethod {
+  public String methodName;
+  public Class<?> returnType;
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/function/PyActorMethod.java]: This should be `Class<T>`. And the `call` method should also return `T`.

File: java/api/src/main/java/org/ray/api/function/PyActorClass.java (hunk 1 above)
Comment (raulchen) [reply java/api/src/main/java/org/ray/api/function/PyActorClass.java]: And document the public fields?

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java [hunk 5]
```diff
@@ -48,26 +53,32 @@
 
   @Test
   public void testCallingPythonFunction() {
-    RayObject res = Ray.callPy(PYTHON_MODULE, "py_func", "hello".getBytes());
+    RayObject res = Ray.call(
+        new PyRemoteFunction(PYTHON_MODULE, "py_func", Object.class),
```
Comment (raulchen) [inline java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: Can we use `byte[].class` here?

File: java/api/src/main/java/org/ray/api/function/PyActorClass.java (hunk 1 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/function/PyActorClass.java]: I think the field name is clear enough. Should I add document?

Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/function/PyActorClass.java]: > `public final` to make it immutable.

I will add public final

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java (hunk 5 above)
Comment (fyrestone) [reply java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: OK. But this field is not used for now.

File: java/api/src/main/java/org/ray/api/runtime/RayRuntime.java [hunk 6]
```diff
@@ -117,40 +140,18 @@
...
-
-  /**
-   * Invoke a remote Python function on an actor.
-   *
-   * @param pyActor A handle to the actor.
-   * @param functionName Name of the actor method.
-   * @param args Arguments of the function.
-   * @return The result object.
-   */
-  RayObject callPyActor(RayPyActor pyActor, String functionName, Object[] args);
```
Comment (kfstorm) [inline java/api/src/main/java/org/ray/api/runtime/RayRuntime.java]: Please avoid reordering methods. This makes it harder to track the changes in commit history and to solve merge conflicts.

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java [hunk 7]
```diff
@@ -2,15 +2,20 @@
 
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableMap;
+
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: No empty lines between imports as Google style suggests.

File: java/api/src/main/java/org/ray/api/function/PyActorMethod.java (hunk 4 above)
Comment (kfstorm) [reply java/api/src/main/java/org/ray/api/function/PyActorMethod.java]: `PyActorMethod<T>` too.

File: java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package org.ray.api.function;
+
+/**
+ * A class for calling Python remote function
+ */
+public class PyRemoteFunction {
```
Comment (kfstorm) [inline java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java]: `PyRemoteFunction <T>`.

File: java/api/src/main/java/org/ray/api/PyActorCall.java [hunk 9]
```diff
@@ -2,40 +2,42 @@
...
+import org.ray.api.function.PyActorMethod;
+
 /**
  * This class provides type-safe interfaces for remote actor calls.
  **/
 @SuppressWarnings({"rawtypes", "unchecked"})
 interface PyActorCall {
 
-  default RayObject call(String functionName) {
+  default RayObject call(PyActorMethod pyActorMethod) {
```
Comment (kfstorm) [inline java/api/src/main/java/org/ray/api/PyActorCall.java]:
```suggestion
  default <R> RayObject<R> call(PyActorMethod<R> pyActorMethod) {
```

`Ray.call` methods too.

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java [hunk 10]
```diff
@@ -48,26 +53,32 @@
...
+    RayObject res = Ray.call(
+        new PyRemoteFunction(PYTHON_MODULE, "py_func", Object.class),
+        "hello".getBytes());
     Assert.assertEquals(res.get(), "Response from Python: hello".getBytes());
   }
 
   @Test
   public void testPythonCallJavaFunction() {
-    RayObject res = Ray.callPy(PYTHON_MODULE, "py_func_call_java_function", "hello".getBytes());
+    RayObject res = Ray.call(
```
Comment (kfstorm) [inline java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: Update all the `RayObject res` to `RayObject<byte[]> res`.

File: java/api/src/main/java/org/ray/api/function/PyActorMethod.java (hunk 4 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/function/PyActorMethod.java]: I added generic type for PyActorMethod and PyRemoteFunction.

File: java/api/src/main/java/org/ray/api/runtime/RayRuntime.java (hunk 6 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/runtime/RayRuntime.java]: Lint says overload functions should be together. So, I reordered the methods.

File: java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java (hunk 8 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java]: Fixed.

File: java/api/src/main/java/org/ray/api/PyActorCall.java (hunk 9 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/PyActorCall.java]: PyCallGenerator.java has updated for generic type.

File: java/api/src/main/java/org/ray/api/RayCall.java [hunk 11]
```diff
@@ -3845,142 +3845,142 @@ public static void call(RayFuncVoid0 f, CallOptions options) {
...
   }
 
-  public static RayObject call(PyRemoteFunction pyRemoteFunction, Object obj0, Object obj1, Object obj2, Object obj3, Object obj4, Object obj5, CallOptions options) {
+  public static <R> RayObject<R> call(PyRemoteFunction<R> pyRemoteFunction, Object obj0, Object obj1, Object obj2, Object obj3, Object obj4, Object obj5, CallOptions options) {
     Object[] args = new Object[]{obj0, obj1, obj2, obj3, obj4, obj5};
     return Ray.internal().call(pyRemoteFunction, args, options);
   }
 
-  public static RayPyActor createActor(PyActorClass pyActorClass) {
+  public static <R> RayPyActor createActor(PyActorClass pyActorClass) {
```
Comment (kfstorm) [inline java/api/src/main/java/org/ray/api/RayCall.java]:
```suggestion
  public static RayPyActor createActor(PyActorClass pyActorClass) {
```

File: java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java (hunk 10 above)
Comment (fyrestone) [reply java/test/src/main/java/org/ray/api/test/CrossLanguageInvocationTest.java]: Fixed.

File: java/api/src/main/java/org/ray/api/RayCall.java (hunk 11 above)
Comment (fyrestone) [reply java/api/src/main/java/org/ray/api/RayCall.java]: I updated RayCallGenerator to generate generic APIs.

File: java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java [hunk 12]
```diff
@@ -0,0 +1,16 @@
+package org.ray.api.function;
+
+/**
+ * A class for calling Python remote function
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java]:
```suggestion
 * A class that represents a Python remote function.
```

Comment (raulchen) [reply java/api/src/main/java/org/ray/api/function/PyRemoteFunction.java]: And document the public fields in these 3 classes.

**Review by raulchen — APPROVED** [review]

**Pull request merged**